### ISSUE TITLE
CLI, Enable / Disable hangs

### ISSUE THREAD

**Issue by retorquere** [Eun/DisableMonitor#27]
I'm trying to use the command line options in a script (on Yosemite), but the screen I'm trying to disable stays enabled, and the call to DisableMonitor doesn't return.

**Comment by Eun** [comment]
Which version are you running?

**Comment by retorquere** [comment]
I should have mentioned that; I'm running 1.9

**Comment by Eun** [comment]
Ok, what have you tried? (In code)
Did you checked #16 out?

**Comment by retorquere** [comment]
I've checked out #16 now:

    ws-180-25:~ emile$ ls -la /Applications/DisableMonitor.app/Contents/MacOS/DisableMonitor
    -rwxr-xr-x@ 1 emile  staff  88704 Jan 31 16:14 /Applications/DisableMonitor.app/Contents/MacOS/DisableMonitor
    ws-180-25:~ emile$ file /Applications/DisableMonitor.app/Contents/MacOS/DisableMonitor
    /Applications/DisableMonitor.app/Contents/MacOS/DisableMonitor: Mach-O 64-bit executable x86_64

I'm running the script I posted in #13 from the command line, not yet from controlplane. I have added DM to my login items, so it starts automatically.

**Comment by Eun** [comment]
So only the disabling does not work? The rest of the command line is working?
E.g. `/Applications/DisableMonitor.app/Contents/MacOS/DisableMonitor --list`

Please be aware that it may take some time until your monitor gets turned off.
So it can be that it appears to hang, but in reality it is waiting.

**Comment by Eun** [comment]
Can you try to disable the monitor via the gui and enable it via cli? Does this work?

**Comment by retorquere** [comment]
The list and help commands work, but disable/enable (long and short form) hang. What is an acceptable wait period?

**Comment by Eun** [comment]
On my MacBook Air it takes sometimes up to 10 seconds, so I assume it could take upto 30 seconds.
In GUI it works without delay?

**Comment by retorquere** [comment]
No delay with the GUI, from the command line I've interrupted it after a minute+

**Comment by retorquere** [comment]
This is on a 2015 Air BTW

**Comment by Eun** [comment]
Ok, need to check this out. If I have access to the model.
You could try some older versions of DisableMonitor though.

If you know Xcode and / or be a programmer you could help out by debugging this.

**Comment by retorquere** [comment]
I can program, but I have no experience with Xcode or Objective-C

**Comment by Eun** [comment]
Well maybe we can meet up for a remote debugging session or something else.
You could try to debug this and see where it hangs. That would help a lot.

**Comment by retorquere** [comment]
Sure, if you can tell me what I need to do, I'm totally new to the Mac platform.

**Comment by Eun** [comment]
So after RemoteDebugging Session it seems that in the CLI the call to `CGDisplayIsOnline` is hanging.

This needs further investigation.

**Comment by retorquere** [comment]
I you need to test again, let me know.

**Comment by retorquere** [comment]
How about having the command line drive the running GUI client via IPC?

**Comment by Eun** [comment]
Could you try if the listing of this app works for you? Or does it hang as well?
https://github.com/nriley/brightness

    brightness -lv

**Comment by afrat** [comment]
Just downloaded this app the other day and everything was working great. Restarted my computer for the first time since installing and now the 'Disable' does not work. Both screens turn off for a second and then both turn back on. The laptop screen is on but not active, meaning I cannot interact with it, but it stays on. I've only tried from the GUI.

**Comment by afrat** [comment]
I'm running version 1.9 by the way.

**Comment by afrat** [comment]
Ok, I found out what was causing it. Apparently it hangs if the monitors aren't mirrored. I enabled 'Mirror Displays' in System Preferences and then disabled via the GUI and it worked. I also tried executing the disable through Terminal before I enabled mirroring and it also was hanging up.

Let me know if I can help debug this in any way, I am not literate in any coding though.

**Comment by retorquere** [comment]
@Eun: if I compile either with `clang -o brightness brightness.c -framework IOKit -framework ApplicationServices -Wl,-U,_CGDisplayModeGetPixelWidth -Wl,-U,_CGDisplayModeGetPixelHeight -mmacosx-version-min=10.6` i get:

    Undefined symbols for architecture x86_64:
      "_main", referenced from:
          start in crt1.10.6.o
    ld: symbol(s) not found for architecture x86_64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)

**Comment by Zouki** [comment]
Hi. Same problem here with OS X Yosemite 10.10.3 on iMac 12,2.
I compiled the code you posted previously and it seems to work.
Output below.

```
iMac:Documents Zouki$ ./displayinfo 
0: 2560x1440 ( active builtin main online )
1: 1920x1080 ( active online )
```

**Comment by Eun** [comment]
Did you get your shell back? Or do you need to press <kbd>CTRL</kbd>+<kbd>C</kbd> to interrupt the process?

**Comment by Zouki** [comment]
I got it back without interrupting.

**Comment by retorquere** [comment]
@Zouki: what command line did you use to compile displayinfo?

**Comment by Zouki** [comment]
I used retorquere's answer so I used
```
clang -o displayinfo displayinfo.c -framework IOKit -framework ApplicationServices -Wl,-U,_CGDisplayModeGetPixelWidth -Wl,-U,_CGDisplayModeGetPixelHeight -mmacosx-version-min=10.6
```
I can try another if necessary.

**Comment by retorquere** [comment]
I hope @Eun has enough info from your run, but it's strange that I get compile errors.

**Comment by Eun** [comment]
@retorquere @Zouki  is using [displayinfo](https://gist.github.com/Eun/bc5bfaea5dc06825c551) not [brightness](https://github.com/nriley/brightness)!

@Zouki  Strange, but ok lets see what @retorquere  reports.

**Comment by retorquere** [comment]
@Eun I tried compiling both the same way

**Comment by Eun** [comment]
@retorquere Please come to Gitter so we can discuss this.

**Comment by retorquere** [comment]
I'm on gitter

**Comment by Eun** [comment]
So apparently before you can use `CGDisplayIsOnline` a call to `CGGetOnlineDisplayList` or `CGGetActiveDisplayList` is mandatory.

**Comment by Eun** [comment]
Fixed in 1.92

**Comment by Gerst20051** [comment]
In order to compile this is the command that I needed to run.

`clang -o displayinfo displayinfo.c -framework ApplicationServices && ./displayinfo`
